WFLY-15836 Add TLS support to JGroups TCP-based transports

rhusar opened the pull request:
Resolves
https://issues.redhat.com/browse/WFLY-15836
https://issues.redhat.com/browse/WFLY-21157

Proposal
https://github.com/wildfly/wildfly-proposals/pull/552

rhusar committed:
Bump jgroups subsystem model to 11.0.0

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
Bump jgroups subsystem schema to 10.0

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
WFLY-15836 Add TLS support to JGroups TCP-based transports
WFLY-21157 Clustering testsuite: missing integration test coverage for TCP_NIO2

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar committed:
Move "WFLY-15836 Add TLS support to JGroups TCP-based transports" feature to Stability.COMMUNITY

This commit will be reverted to promote to Stability.DEFAULT

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

rhusar commented:
@darranl Given that this is integrating with Elytron, could you or someone from your team have a look too? Thanks!

rhusar commented on clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/SocketTransportResourceDefinitionRegistrar.java:
```diff
@@ -98,16 +134,35 @@ public T createProtocol(ChannelFactoryConfiguration stackConfiguration) {
+
+                    @Override
+                    public Optional<TLSConfiguration> getSSLConfiguration() {
+                        if (serverSSLContext.get() != null && clientSSLContext.get() != null) {
```
Done.

rhusar commented on clustering/jgroups/extension/src/test/java/org/jboss/as/clustering/jgroups/subsystem/JGroupsSubsystemTestCase.java:
```diff
@@ -45,13 +47,13 @@ public static Iterable<JGroupsSubsystemSchema> parameters() {
 
     public JGroupsSubsystemTestCase(JGroupsSubsystemSchema schema) {
-        super(JGroupsSubsystemResourceDefinitionRegistrar.REGISTRATION.getName(), new JGroupsExtension(), schema, JGroupsSubsystemSchema.CURRENT);
+        super(JGroupsSubsystemResourceDefinitionRegistrar.REGISTRATION.getName(), new JGroupsExtension(), schema, JGroupsSubsystemSchema.CURRENT.stream().filter(s -> s.getStability().equals(Stability.DEFAULT)).findFirst().get());
```
Fixed.

rhusar commented on clustering/jgroups/extension/src/test/java/org/jboss/as/clustering/jgroups/subsystem/OperationTestCaseBase.java:
```diff
@@ -25,7 +26,7 @@
 public class OperationTestCaseBase extends AbstractSubsystemTest {
 
-    static final String SUBSYSTEM_XML_FILE = String.format("jgroups-%d.%d.xml", JGroupsSubsystemSchema.CURRENT.getVersion().major(), JGroupsSubsystemSchema.CURRENT.getVersion().minor());
+    static final String SUBSYSTEM_XML_FILE = String.format("jgroups-%d.%d.xml", JGroupsSubsystemSchema.CURRENT.stream().filter(s -> s.getStability().equals(Stability.DEFAULT)).findFirst().get().getVersion().major(), JGroupsSubsystemSchema.CURRENT.stream().filter(s -> s.getStability().equals(Stability.DEFAULT)).findFirst().get().getVersion().minor());
```
Done.

rhusar commented on clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/SocketTransportResourceDefinitionRegistrar.java:
```diff
@@ -66,23 +89,36 @@ public ResourceOperationRuntimeHandler getParentRuntimeHandler() {
                         T transport = super.createProtocol(stackConfiguration);
-                        if (clientSocketBinding != null) {
-                            InetSocketAddress clientSocketAddress = clientSocketBinding.getSocketAddress();
+                        if (clientSocketBinding.get() != null) {
```
Done.

rhusar commented on clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/TLSManagedSocketFactory.java:
```diff
@@ -0,0 +1,71 @@
+    public ServerSocket createServerSocket(String name) throws IOException {
+        SSLServerSocketFactory factory = this.serverSSLContext.getServerSocketFactory();
+        ServerSocket serverSocket = factory.createServerSocket();
+        Closeable registeredSocket = this.manager.getNamedRegistry().registerSocket(name, serverSocket);
```
Done please rereview this part.

pferraro reviewed (approved):
Remaining suggestions are optional.

rhusar commented:
> @rhusar I am going to include this in my bundle for testing, this looks close - where would you like the feature team to discuss this satisifes the community requirements?

Started here: [#wildfly-developers > WFLY-15836 Add TLS support to JGroups TCP-based transports](https://wildfly.zulipchat.com/#narrow/channel/174184-wildfly-developers/topic/WFLY-15836.20Add.20TLS.20support.20to.20JGroups.20TCP-based.20transports/with/562943945)

pferraro commented on clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/SocketTransportResourceDefinitionRegistrar.java:
```diff
@@ -50,6 +56,23 @@ public PathElement getPathElement() {
             .setAccessConstraints(SensitiveTargetAccessConstraintDefinition.SOCKET_BINDING_REF)
             .build();
 
+    private static final String CLIENT_CONTEXT_NAME = "client-context";
```
Can we change this to: "client-ssl-context"?

rhusar replied:
We might need to, because these names are appearing in the console.

Context - [#wildfly-developers > Console issues in JGroups subsystem](https://wildfly.zulipchat.com/#narrow/channel/174184-wildfly-developers/topic/Console.20issues.20in.20JGroups.20subsystem/with/562965689)

pferraro commented on clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/JGroupsSubsystemSchema.java:
```diff
@@ -234,6 +240,15 @@ public ModelNode apply(ModelNode operation) {
         }
+
+        if (this.since(JGroupsSubsystemSchema.VERSION_9_0_COMMUNITY)) {
+            ResourceXMLElement ssl = this.factory.element(this.factory.resolve("ssl"))
```
I think the following would be better:
```xml
<ssl-context client="..." server="..."/>
```

pferraro reviewed (changes requested):
Some suggestions for attribute model/xml name changes.

rhusar committed:
WFLY-15836 Simplify JGroups transport SSL context configuration to client-ssl-context/server-ssl-context
- Rename XML element from 'ssl' to 'ssl-context' and attributes to 'client' and 'server' (shorter names within ssl-context element)
- Rename model attributes to 'client-ssl-context' and 'server-ssl-context'
- Update schema, documentation, tests, and test setup accordingly

Signed-off-by: Radoslav Husar <radosoft@gmail.com>

pferraro reviewed (approved):
Nice work!
